### PR TITLE
HADOOP-18129: Change URI to String in INodeLink to reduce memory footprint of ViewFileSystem

### PR DESCRIPTION
**Pull request by abhishekdas99**
https://issues.apache.org/jira/browse/HADOOP-18129

### For code changes:

- [ ] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [ ] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   1m  9s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  2s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  12m 50s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  26m 47s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  25m 58s |  |  trunk passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  compile  |  23m 35s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  checkstyle  |   4m  5s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 13s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   2m  3s |  |  trunk passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   2m 22s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   3m 22s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  26m 23s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 25s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 33s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  28m 59s |  |  the patch passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javac  |  28m 59s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  26m 28s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  javac  |  26m 28s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   4m 20s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3996/1/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 1 new + 179 unchanged - 1 fixed = 180 total (was 180)  |
| +1 :green_heart: |  mvnsite  |   2m 14s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m 44s |  |  the patch passed with JDK Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04  |
| +1 :green_heart: |  javadoc  |   2m 19s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   3m 59s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  24m  4s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  31m  4s |  |  hadoop-common in the patch passed.  |
| -1 :x: |  unit  |  47m 58s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs-nfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3996/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs-nfs.txt) |  hadoop-hdfs-nfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 48s |  |  The patch does not generate ASF License warnings.  |
|  |   | 310m 30s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.nfs.nfs3.TestNfs3HttpServer |
|   | hadoop.hdfs.nfs.nfs3.TestClientAccessPrivilege |
|   | hadoop.hdfs.nfs.nfs3.TestViewfsWithNfs3 |
|   | hadoop.hdfs.nfs.nfs3.TestRpcProgramNfs3 |
|   | hadoop.hdfs.nfs.nfs3.TestReaddir |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3996/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3996 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux 29f0024893b6 4.15.0-163-generic #171-Ubuntu SMP Fri Nov 5 11:55:11 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / a75520f0ec1fff39c4b58402bf4f331fe4dc9737 |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.13+8-Ubuntu-0ubuntu1.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3996/1/testReport/ |
| Max. process+thread count | 3143 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs-nfs U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3996/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java [hunk 1]
```diff
@@ -216,19 +216,19 @@ public boolean equals(Object obj) {
...
+      targetFileSystemPaths = targetFs;
     }
 
     public Path getMountedOnPath() {
       return mountedOnPath;
     }
 
-    public URI[] getTargetFileSystemURIs() {
-      return targetFileSystemURIs;
+    public String[] getTargetFileSystemPaths() {
```
Comment (ibuenros) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java:238]: This is a public method, can we keep the old signature? We can transform from string to URI within the method itself.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/InodeTree.java [hunk 2]
```diff
@@ -294,11 +294,11 @@ void addLink(final String pathComponent, final INodeLink<T> link)
...
     INodeLink(final String pathToNode, final UserGroupInformation aUgi,
         Function<URI, T> createFileSystemMethod,
-        final URI aTargetDirLink) {
+        final String aTargetDirLink) throws URISyntaxException {
       super(pathToNode, aUgi);
       targetFileSystem = null;
-      targetDirLinkList = new URI[1];
-      targetDirLinkList[0] = aTargetDirLink;
+      targetDirLinkList = new String[1];
+      targetDirLinkList[0] = new URI(aTargetDirLink).toString();
```
Comment (ibuenros) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/InodeTree.java:301]: What is the purpose of creating a URI and converting back to string?

**Review by ibuenros — CHANGES_REQUESTED** [review]
Can you also comment on performance implications of this change? Specifically `ViewFileSystem.getTargetFileSystemPaths` may now require parsing strings into URIs on each usage, where does this impose a performance penalty?

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/InodeTree.java (hunk 2 above)
Comment (abhishekdas99) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/InodeTree.java:301]: In the previous implementation, we were creating the URI object from the target file system path. If the path is not a valid URI, the filesystem initialization should have failed. As we are not keeping the URI object anymore, this ensures we are not dealing with an invalid URI and new URI object validates the target filesystem path.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java (hunk 1 above)
Comment (abhishekdas99) [reply hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java:238]: Added getTargetFileSystemURIs where we are returning URI[]. Also added method to return targetFileSystem Paths.

**Comment by abhishekdas99** [comment]
> Can you also comment on performance implications of this change? Specifically `ViewFileSystem.getTargetFileSystemPaths` may now require parsing strings into URIs on each usage, where does this impose a performance penalty?

The performance impact of this change where we are creating the URI object when its needed as well as creating a dummy URI object to check the validity of the path.
I created 100000 URI objects in less than a second ( It took 700 msec to be precise).

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 50s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  12m 31s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  25m 54s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  24m 27s |  |  trunk passed with JDK Ubuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  compile  |  21m  0s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  checkstyle  |   3m 57s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 14s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m 43s |  |  trunk passed with JDK Ubuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   2m  6s |  |  trunk passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   3m 22s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  23m 43s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 24s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 19s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  26m 25s |  |  the patch passed with JDK Ubuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javac  |  26m 25s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  24m 42s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  javac  |  24m 42s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   4m  1s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3996/2/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 1 new + 179 unchanged - 1 fixed = 180 total (was 180)  |
| +1 :green_heart: |  mvnsite  |   2m 29s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m 43s |  |  the patch passed with JDK Ubuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04  |
| +1 :green_heart: |  javadoc  |   2m 12s |  |  the patch passed with JDK Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07  |
| +1 :green_heart: |  spotbugs  |   4m 51s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  25m 33s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  18m 51s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  unit  |   3m 24s |  |  hadoop-hdfs-nfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 51s |  |  The patch does not generate ASF License warnings.  |
|  |   | 241m 51s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3996/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/3996 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell |
| uname | Linux c1e481352f5d 4.15.0-163-generic #171-Ubuntu SMP Fri Nov 5 11:55:11 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / d69c720598a2e5d8eadb145bfc3b8b47b6d16bd9 |
| Default Java | Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.14+9-Ubuntu-0ubuntu2.20.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_312-8u312-b07-0ubuntu1~20.04-b07 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3996/2/testReport/ |
| Max. process+thread count | 1286 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-hdfs-project/hadoop-hdfs-nfs U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-3996/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

**Comment by omalley** [comment]
What do the numbers look like for the memory usage before and after this change?

**Comment by abhishekdas99** [comment]
> What do the numbers look like for the memory usage before and after this change?

For 40k mount points, the size of ViewFileSystem reduced from 62MB to 28MB.

**Review by omalley — APPROVED** [review]
LGTM +1

**Pull request closed without merging**